crates/apollo-router-core/src/query_planner/model.rs

```diff
@@ -3,15 +3,54 @@
 //!
 //! QueryPlans are a set of operations that describe how a federated query is processed.
 
+use std::collections::HashMap;
+
 use crate::prelude::graphql::*;
+use apollo_parser::ast;
 use serde::{Deserialize, Serialize};
 
+/// The root query plan container.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[serde(tag = "kind")]
+pub struct JsQueryPlan {
+    /// The hierarchical nodes that make up the query plan
+    pub node: Option<PlanNode>,
+}
+
+/*FIXME: I get a panic when calling JS if I add the fields to QueryPlan:
+thread 'tokio-runtime-worker' panicked at 'unable to invoke var _a;
+Object.defineProperty(exports, "__esModule", { value: true });
+const planResult = bridge.plan(schemaString, queryString, operationName);
+if (((_a = planResult.errors) === null || _a === void 0 ? void 0 : _a.length) > 0) {
+    done({ Err: planResult.errors });
+}
+else {
+    done({ Ok: planResult.data });
+}
+//# sourceMappingURL=do_plan.js.map in JavaScript runtime
+ error:
+ Error: Error parsing args: serde_v8 error: ExpectedArray
+    at unwrapOpResult (deno:core/core.js:99:13)
+    at Object.opSync (deno:core/core.js:113:12)
+    at done (<init>:7:15)
+    at do_plan:8:5', /home/geal/.cargo/git/checkouts/federation-320f8bad94ab52a0/1ffecef/router-bridge/src/js.rs:105:13
+note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
+
+*/
 /// The root query plan container.
 #[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
 #[serde(tag = "kind")]
 pub struct QueryPlan {
     /// The hierarchical nodes that make up the query plan
     pub node: Option<PlanNode>,
+    // list of operations to apply on the final response
+    #[serde(default)]
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    pub operations: Vec<Operation>,
+    // list of fragments to apply on the final response
+    #[serde(default)]
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    pub fragments: HashMap<String, Fragment>,
```

Comment:
Correct me if I'm wrong:

Here you are adding new fields in the QueryPlan that will serve as "cache" for information that come from the query itself.

You are also re-using the existing objects (InlineFragment, Selection) and creating new ones (FragmentSpread) to transform some of the information that come from apollo-parser.

SO. I wonder:
1. Is this the right place for caching? Obviously QueryPlan is already cached and exist per-query so it kinda make sense. Though it's supposed to be the query's plan object and not the query's object. My point is: maybe this should be renamed to `QueryMetadata`?
2. Is re-using the object a good idea? I understand this file is called model.rs but it was supposed to be the model of the query planner (hence the parent module is called query_planner). When I ask the query planner, can I receive FragmentSpread? The code before didn't allow that.

    I have the impression that you should keep the 2 models separated: one come from parsing the query itself, the other one come from parsing the plan from the query planner.

Reply:
I'm not sure it's the right place, but it feels coherent to have it close to the query plan. `QueryMetadata` can work, with a field for the plan and others for operations and fragments?

I'm not sure if we can receive `FragmentSpread `, we should test for that. I found it weird that this particular structure was not there while `InlineFragment` was defined.

I'm ok with separating them, especially if we're risking collisions with what the query planner is returning